Intermediate tour: Libraries and APIs chapter (#4743)
JetBrains/kotlin-web-site: docs/topics/tour/kotlin-tour-intermediate-libraries-and-apis.md

[//]: # (title: Intermediate: Libraries and APIs)

To get the most out of Kotlin, use existing libraries and APIs so you can spend more time coding and less time 
reinventing the wheel.

Libraries distribute reusable code that simplifies common tasks. Within libraries, there are packages and objects that
group related classes, functions, and utilities. Libraries expose APIs (Application Programming Interfaces) as a set of 
functions, classes, or properties that developers can use in their code.

Let's explore what's possible with Kotlin.

## The standard library

Kotlin has a standard library that provides essential types, functions, collections, and utilities to make your code 
concise and expressive. A large portion of the standard library (everything in the [`kotlin` package](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/)) is readily available

> in any Kotlin file without you having to explicitly import it:

Reviewer: maybe a bit shorter:
> without the need to import it explicitly:

```kotlin
fun main() {
    val text = "emosewa si niltoK"
    
   // Use the reversed() function from the standard library
    val reversedText = text.reversed()

    // Use the print() function from the standard library
    print(reversedText)
    // Kotlin is awesome
}
```
{kotlin-runnable="true" id="kotlin-tour-libraries-stdlib"}

> But there are some parts of the standard library that require you to import them before using them in your code. 

Reviewer: ```suggestion
However, some parts of the standard library require an import before you can use them in your code. 
```

For example, if you want to use the standard library's time measurement features, you need to import the [`kotlin.time` package](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.time/).

> At the top of your file, or before where you use it in your code, add the `import` keyword followed by the package that you need:

Reviewer: Imports are always at the top of the file or I didn't understand the meaning

or maybe you meant "Before using it in your code, at the top of the file, add the `import` keyword followed by the package that you need:"

Author: You're right. Fixed!

```kotlin
import kotlin.time.*
```

The asterisk `*` is a wildcard import that tells Kotlin to import everything within the package. You can't use the 

> asterisk `*` with companion objects.

Reviewer: ```suggestion
asterisk `*` with companion objects too.
```

Author: I'm not sure I understand why you think the "too" should be here 🤔 Can you explain a bit?

For example:

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes

fun main() {
    val thirtyMinutes: Duration = 30.minutes
    val halfHour: Duration = 0.5.hours
    println(thirtyMinutes == halfHour)
    // true
}
```
{kotlin-runnable="true" id="kotlin-tour-libraries-time"}

This example:

* Imports the `Duration` class and the `hours` and `minutes` extension properties from its companion object.
* Uses the `minutes` property to convert `30` into a `Duration` of 30 minutes.
* Uses the `hours` property to convert `0.5` into a `Duration` of 30 minutes.
* Checks if both durations are equal and prints the result.

### Search before you build

Before you decide to write your own code, check the standard library to see if what you're looking for already exists. 

> Here's a list of areas where the standard library already provides a number of classes, functions, and properties for your use:

Reviewer: ```suggestion
Here's a list of areas where the standard library already provides a number of classes, functions, and properties for you:
```

* [Managing collections](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/)
* [Sequences](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.sequences/)
* [String manipulation](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.text/)

> To learn more about what else is inside the standard library, explore the standard library [API reference](https://kotlinlang.org/api/core/kotlin-stdlib/).

Reviewer: ```suggestion
To learn more about what else is in the standard library, explore its [API reference](https://kotlinlang.org/api/core/kotlin-stdlib/).
```

## Kotlin libraries

The standard library covers many common use cases, but there are some that it doesn't address. Fortunately, the Kotlin
team and the rest of the community have developed a wide range of libraries to complement the standard library. For example,
[`kotlinx-datetime`](https://kotlinlang.org/api/kotlinx-datetime/) helps you manage time across different platforms.

You can find useful libraries on our [search platform](https://klibs.io/). To use them, you need to take extra steps,
like adding a dependency or plugin. Each library has a GitHub repository with instructions on how to include
it in your Kotlin projects.

Once you add the library, you can import any package within it. Here's an example of how to import the `kotlinx-datetime`
package to find the current time in New York: 

```kotlin
import kotlinx.datetime.*

fun main() {

>     val now = Clock.System.now()  // Get current instant

Reviewer: ```suggestion
    val now = Clock.System.now() // Get current instant
```

```kotlin
@OptIn(ExperimentalUnsignedTypes::class)
fun main() {
    // Create an unsigned integer array
    val unsignedArray: UIntArray = uintArrayOf(1u, 2u, 3u, 4u, 5u)

    // Modify an element
    unsignedArray[2] = 42u
    println("Updated array: ${unsignedArray.joinToString()}")
    // Updated array: 1, 2, 42, 4, 5
}
```
{kotlin-runnable="true" id="kotlin-tour-libraries-apis"}

This is the easiest way to opt in, but there are other ways. To learn more, see [Opt-in requirements](opt-in-requirements.md).

## Libraries and APIs practice

### Exercise 1 {initial-collapse-state="collapsed" collapsible="true" id="libraries-exercise-1"}

You are developing a financial application that helps users calculate the future value of their investments. The formula
to calculate compound interest is:

<math>A = P \times (1 + \displaystyle\frac{r}{n})^{nt}</math>

Where:

* `A` is the amount of money accumulated after interest (principal + interest).
* `P` is the principal amount (the initial investment).
* `r` is the annual interest rate (decimal).
* `n` is the number of times interest is compounded per year.

> * `t` is the time the money is invested for, in years.

Reviewer: ```suggestion
* `t` is the time the money is invested for (in years).
```

    // The accumulated amount is: 1282.0372317085844
}

```
{validate="false" kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-tour-libraries-exercise-1"}

|---|---|
```kotlin
import kotlin.math.*

fun calculateCompoundInterest(P: Double, r: Double, n: Int, t: Int): Double {
    return P * (1 + r / n).pow(n * t)
}

fun main() {
    val principal = 1000.0
    val rate = 0.05
    val timesCompounded = 4
    val years = 5
    val amount = calculateCompoundInterest(principal, rate, timesCompounded, years)
    println("The accumulated amount is: $amount")
    // The accumulated amount is: 1282.0372317085844
}
```
{initial-collapse-state="collapsed" collapsible="true" collapsed-title="Example solution" id="kotlin-tour-libraries-solution-1"}

### Exercise 2 {initial-collapse-state="collapsed" collapsible="true" id="libraries-exercise-2"}

You want to measure the time it takes to perform multiple data processing tasks in your program. Update the code
to add the correct import statements and functions from the [`kotlin.time`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.time/) package:

|---|---|

```kotlin
// write your code here

fun main() {

>     val timeTaken = // write your code here {

Author: They need to search in the `kotlin.time` package and find the `measureTime()` function.

>     val timeTaken = // write your code here {

Reviewer: Oh, I see! then it should be `val timeTaken = /* write your code here */ {` so that opening bracket is not commented out
